### DISCUSSION TITLE
Tons of warning and no error during make PythonAPI, and the carla/dist/ directory is not created

### DISCUSSION THREAD

**Discussion by GarfieldGa** [carla-simulator/carla#4248] [Q&A] [answered]
I'm currently setting up the carla environment, and have successfully finished make PythonAPI and `make launch` with no errors. I can now see the map and move around, but when I tried to run the spawn_npc.py, script it's not working. I found that even though there is no error during the previous steps, there are tons of warnings and the expected PythonAPI/carla/dist directory is not created.

Then I tried the following command recommended by some other user:
```
cd PythonAPI/carla
python setup.py install

```
And there are errors coming as:
![image](https://user-images.githubusercontent.com/43100023/119557890-60022800-bd66-11eb-8e9e-c6f4c914b4a4.png)
I'm runing the prompt as admin and the error translates as parameter type mismatch (can't convert std::hash<carla::road::element::Waypoint>::result_type to size_t &)

I'm using:
win10 for workstation (64bit)
16GB RAM
Nvidia Quadro P3200
Intel Xeon E-2176M

Python 3.7.8
UE 4.24.3
carla 0.9.11

**Comment by corkyw10** [comment]
Hi @GarfieldGa, sometimes the `make PythonAPI` command in Windows can finish with a success message but actually wasn't successful. If the `.egg` file was not created in `PythonAPI/carla/dist` then it's likely something when wrong with that part of the build. I suggest starting with a clean installation and when you finish with `make PythonAPI`, check for the  `PythonAPI/carla/dist/<carla_egg_file>` and if it's not there check if there were any errors just before the final success message of the command. You can post any errors here and I will try to help.

**Reply by GarfieldGa** [reply]
Hi, thank you for your response, and I checked that I didn't face any error during make PythonAPI. A strange thing is: not a single file seems to have output file during make PythonAPI i.e., there is no dist directory created and I only got warnings...

Below is example of message I get, they are pretty similar in content:
`C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Idependencies/include -IC:\Users\Garfield_hant\AppData\Local\Programs\Python\Python37\include -IC:\Users\Garfield_hant\AppData\Local\Programs\Python\Python37\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\Include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\Include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\cppwinrt" /EHsc /Tpdependencies/include/carla\trafficmanager\ALSM.cpp /Fobuild\temp.win-amd64-3.7\Release\dependencies/include/carla\trafficmanager\ALSM.obj /experimental:external /external:I dependencies/include/system /DBOOST_ALL_NO_LIB /DBOOST_PYTHON_STATIC_LIB /DBOOST_ERROR_CODE_HEADER_ONLY /D_WIN32_WINNT=0x0600 /DHAVE_SNPRINTF /DLIBCARLA_WITH_PYTHON_SUPPORT -DLIBCARLA_IMAGE_WITH_PNG_SUPPORT=true`

**Comment by GarfieldGa** [comment, marked as answer]
I think I now get carla working on my windows:
![ed3b231e0104744428cb4df59a5fb94](https://user-images.githubusercontent.com/43100023/119752306-637ad980-be62-11eb-9165-9584b5505b3b.png)

**The problem I was facing is:**
I got two different versions of python on my machine, one is 32-bit and the other is 64-bit.
And the `python` command in my cmd refers to the 32-bit version while `py` command is the 64-bit version.
`make PythonAPI` seems to call `python` command as default internally (in setup.py) so that `make PythonAPI` never succeed.
And when I try `py setup.py install`, the error reveals as C2664, can't perform type conversion.
And for some reason, python3.7.9 seems not working with carla? I keep getting error: Microsoft Visual C++ 14.0 is required..

Note:
This error I face is actually posted by someone else and if you are facing the same error, add explicit casting in Waypoint.cpp so that you can see the true error when recompiling using `python setup.py install`:
```
#include "carla/road/element/Waypoint.h"

#include <boost/container_hash/hash.hpp>

namespace std {

  using WaypointHash = hash<carla::road::element::Waypoint>;

  WaypointHash::result_type WaypointHash::operator()(const argument_type &waypoint) const {
    WaypointHash::result_type seed = 0u;
    boost::hash_combine((size_t &) seed, waypoint.road_id);
    boost::hash_combine((size_t &) seed, waypoint.section_id);
    boost::hash_combine((size_t &) seed, waypoint.lane_id);
    boost::hash_combine((size_t &) seed, static_cast<float>(std::floor(waypoint.s * 200.0)));
    return seed;
  }

} // namespace std
```
Using explicit type conversion, you should be able to get something like:
LNK1112: module machine type 'x64' conflicts with target machine type 'X86'

As for the printout, I think these should be the normal printout. And there should be warnings around, about possible loss of data during type convertion (since size_t is dependent on system arch). This means the printout of `make PythonAPI` doesn't report anything special when the compilation fails.
